core/TypePtr.cc

```diff
@@ -91,4 +91,325 @@ bool TypePtr::isBottom() const {
     return t != nullptr && t->symbol == Symbols::bottom();
 }
 
+int TypePtr::kind() const {
+    switch (tag()) {
+        case Tag::AppliedType:
+            return 1;
+        case Tag::BlamedUntyped:
+        case Tag::UnresolvedAppliedType:
+        case Tag::UnresolvedClassType:
+        case Tag::ClassType:
+            return 2;
+        case Tag::LiteralType:
+            return 3;
+        case Tag::ShapeType:
+            return 4;
+        case Tag::TupleType:
+            return 5;
+        case Tag::LambdaParam:
+        case Tag::SelfTypeParam:
+            return 6;
+        case Tag::MetaType:
+            return 7;
+        case Tag::TypeVar:
+            return 8;
+        case Tag::AliasType:
+            return 9;
+        case Tag::OrType:
+            return 10;
+        case Tag::AndType:
+            return 11;
+        case Tag::SelfType:
+            return 12;
+    }
+}
+
+std::string TypePtr::typeName() const {
+    switch (tag()) {
+        case Tag::AppliedType:
+            return "AppliedType";
+        case Tag::BlamedUntyped:
+            return "BlamedUntyped";
+        case Tag::UnresolvedAppliedType:
+            return "UnresolvedAppliedType";
+        case Tag::UnresolvedClassType:
+            return "UnresolvedClassType";
+        case Tag::ClassType:
+            return "ClassType";
+        case Tag::LiteralType:
+            return "LiteralType";
+        case Tag::ShapeType:
+            return "ShapeType";
+        case Tag::TupleType:
+            return "TupleType";
+        case Tag::LambdaParam:
+            return "LambdaParam";
+        case Tag::SelfTypeParam:
+            return "SelfTypeParam";
+        case Tag::MetaType:
+            return "MetaType";
+        case Tag::TypeVar:
+            return "TypeVar";
+        case Tag::AliasType:
+            return "AliasType";
+        case Tag::OrType:
+            return "OrType";
+        case Tag::AndType:
+            return "AndType";
+        case Tag::SelfType:
+            return "SelfType";
+    }
+}
+
+bool TypePtr::isFullyDefined() const {
+    switch (tag()) {
```

Comment:
This version of `isFullyDefined` is really easy to reason about, with most of the cases all in one place 👍